cdap-ui/app/cdap/components/FastAction/SetPreferenceAction/index.js

```diff
@@ -0,0 +1,442 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React, {Component, PropTypes} from 'react';
+import FastActionButton from '../FastActionButton';
+import isObject from 'lodash/isObject';
+import upperFirst from 'lodash/upperFirst';
+import debounce from 'lodash/debounce';
+import orderBy from 'lodash/orderBy';
+import T from 'i18n-react';
+import shortid from 'shortid';
+import { Modal, Tooltip, ModalHeader, ModalBody } from 'reactstrap';
+import myPreferenceApi from 'api/preference';
+import {convertProgramToApi} from 'services/program-api-converter';
+import KeyValuePairs from 'components/KeyValuePairs';
+import NamespaceStore from 'services/NamespaceStore';
+require('./SetPreferenceAction.less');
+
+export default class SetPreferenceAction extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      modal: false,
+      loading: false,
+      errorMessage: '',
+      extendedMessage: '',
+      savedMessage: null,
+      error: '',
+      keyValues: {'pairs': [{'key':'', 'value':'', 'uniqueId': ''}]},
+      inheritedPreferences: [],
+      sortedByColumn: 'key',
+      sortOrder: 'asc',
+    };
+
+    this.params = {
+      namespace: NamespaceStore.getState().selectedNamespace,
+      appId: this.props.entity.id,
+    };
+
+    if (this.props.entity.type == 'program') {
+      this.params.appId = this.props.entity.applicationId;
+      this.params.programId = this.props.entity.id;
+      this.params.programType = convertProgramToApi(this.props.entity.programType);
+    }
+
+    this.debouncedSetPreferences = debounce(this.debounceSetPreferences, 1000);
+    this.eventText = '';
+    this.toggleModal = this.toggleModal.bind(this);
+    this.onKeyValueChange = this.onKeyValueChange.bind(this);
+    this.toggleTooltip = this.toggleTooltip.bind(this);
+  }
+
+  toggleTooltip() {
+    this.setState({ tooltipOpen : !this.state.tooltipOpen });
+  }
+  toggleModal(event) {
+    this.setState({
+      modal: !this.state.modal,
+      loading: false,
+      errorMessage: '',
+      extendedMessage: '',
+      savedMessage: null,
+      textarea: false,
+      textInput: '',
+      droppedFile: {}
+    });
+    if (event) {
```

Comment:
usually if you need to stop event propagation, you should do this as first thing in the function. `this.setState()` will trigger `render()` and evaluate those synchronously